libs/tlv/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+            fn encode<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+                crate::value::encode_tagged(Tag::U8(0x01), &self.a, writer)?;
+                crate::value::encode_tagged(Tag::U8(0x02), &self.b, writer)?;
+                Ok(())
+            }
+
+            fn encoded_length(&self) -> std::io::Result<crate::Length> {
+                let mut length = Length::new(0);
+                length += crate::value::length_tagged(Tag::U8(0x01), &self.a)?;
+                length += crate::value::length_tagged(Tag::U8(0x01), &self.b)?;
+                Ok(length)
+            }
+        }
+
+        let value = Test { a: 0x99, b: 0xabcd };
+        let encoded = [0x01, 0x01, 0x99, 0x02, 0x02, 0xab, 0xcd];
+
+        assert_eq!(to_vec(&value).unwrap(), encoded);
+        assert_eq!(value.encoded_length().unwrap(), Length::new(7));
+    }
+
+    #[test]
+    fn test_decode_codegen() {
+        struct Test {
+            a: u8,
+            b: u16,
+        }
+
+        // This is implemented in a way that is conducive to code generation.
+        impl Decode for Test {
+            fn decode<R: std::io::Read>(reader: &mut R) -> std::io::Result<(Self, usize)> {
+                let tlv_decode_read: usize = 0;
+                let (a, tlv_decode_read) = match crate::value::decode_tagged(Tag::U8(0x01), reader)?
+                {
+                    (a, read) => (a, tlv_decode_read + read),
+                };
+                let (b, tlv_decode_read) = match crate::value::decode_tagged(Tag::U8(0x02), reader)?
+                {
+                    (b, read) => (b, tlv_decode_read + read),
+                };
+                Ok((Test { a, b }, tlv_decode_read))
+            }
+        }
+
+        let data = vec![0x01, 0x01, 0x99, 0x02, 0x02, 0xab, 0xcd];
+        let decoded: Test = from_slice(&data).unwrap();
+        assert_eq!(decoded.a, 0x99);
+        assert_eq!(decoded.b, 0xabcd);
+    }
+
+    #[test]
+    fn test_medium_length() {
+        let tag = Tag::U8(0x01);
+        let value = vec![0xbe; 0x81];
+        let encoded = to_vec_tagged(tag, &value).unwrap();
+
+        //                        tag    MED   len   data …
+        assert_eq!(encoded[0..4], [0x01, 0x81, 0x81, 0xbe]);
+
+        let decoded: Vec<u8> = from_slice_tagged(tag, &encoded).unwrap();
+        assert_eq!(decoded, value);
+    }
+
+    #[test]
+    fn test_long_length() {
+        let tag = Tag::U8(0x01);
+        let value = vec![0xbe; 0x100];
+        let encoded = to_vec_tagged(tag, &value).unwrap();
+
+        //                        tag    LONG  len0  len1  data …
+        assert_eq!(encoded[0..5], [0x01, 0x82, 0x01, 0x00, 0xbe]);
+
+        let decoded: Vec<u8> = from_slice_tagged(tag, &encoded).unwrap();
+        assert_eq!(decoded, value);
+    }
+
+    #[test]
+    fn test_boolean() {
+        assert_eq!(to_vec(true).unwrap(), [0x01]);
+        assert_eq!(to_vec(false).unwrap(), [0x00]);
+    }
+
+    proptest! {
```

Comment:
The test functions in this block will be called with automatic random values for each of the arguments, similar to `fast-check` in JS but with less boilerplate.